File: basic-to-master/data-structure/9012.py
# sol 1)
# t = int(input())
# for _ in range(t):
#     s = []
#     for p in input():
#         if p == '(':
#             s.append(p)
#         elif len(s) > 0 and p == ')':
#             if s.pop() != '(':
#                 print('NO')
#                 break
#         elif len(s) == 0:
#             print('NO')
#             break
#     else:
#         if len(s) > 0:
#             print('NO')
#         else:
#             print('YES')

# sol 2)
for _ in range(int(input())):
    stk = []
    isVPS = True
    for ch in input():
        if ch == '(':
            stk.append(ch)
        else:
            if stk:
                stk.pop()
            else:
                isVPS = False
                break

    if stk:
        isVPS = False

    print('YES' if isVPS else 'NO')

File: cpp/바킹독_알고리즘/스택의 활용(수식의 괄호 쌍)/9012.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    int tc;
    cin >> tc;
    while (tc--) {
        string s;
        cin >> s;
        stack<char> stk;
        bool isValid = true;
        for (char ch : s) {
            if (ch == '(') stk.push(ch);
            else if (!stk.empty()) stk.pop();
            else {
                isValid = false;
                break;
            }
        }

        if (!stk.empty()) isValid = false;

        if (isValid) cout << "YES\n";
        else cout << "NO\n";
    }
}